### PR TITLE
Add full autocompletion capabilities to language server

### PR DESCRIPTION
**Pull request by brandonlangley**
Add parameter, block, and value autocompletion to language server

NEAMS Workbench demonstrating some capabilities on this branch:

https://github.com/idaholab/moose/assets/21372118/253792ce-883c-480b-b68c-176586d23eed

refs #22766

### CODE REVIEW

**Comment by brandonlangley** [comment]
@loganharbour @dschwen -

This adds full autocompletion capabilities to the MOOSE language server.

It is all working really well and is super snappy in the NEAMS Workbench.

I had to make two minor update in the Workbench code to:

 - use LSP documentation field rather than detail field for descriptions

 - truncate long descriptions in the autocomplete list for cleaner display

These changes haven't made a Workbench release yet, but they will soon.

This uses syntax paths and object types (or defaults) of the cursor position.

And it basically handles autocomplete in three main high level categories:

 - input parameters

   - starting with parser global parameters (currently active / inactive)

   - plus all action parameters for the object context path of the cursor

   - plus all object parameters for object type if it is valid for syntax

   - building an input default for each parameter and using descriptions

   - indicating in description when a parameter is required in the input

   - removing requirement via object_params_set_by_action as needed

   - filtering parameters that are deprecated, private, or already exist

 - blocks / subblocks

   - top level blocks when at document root level outside of all blocks

   - all '*' subblocks and named subblocks from within any block context

   - inserting opening and closing brace syntax of blocks on completion

 - parameter values

   - adding existing subblock children names for active / inactive values

   - adding 'true' and 'false' string values if parameter type is boolean

   - adding valid enums and descriptions if parameter is an enum type

   - adding all verified object names if (non-enum) parameter is 'type'

   - adding input lookup values if parameter type has associated syntax

     - i.e., names of Functions, Variables, Materials, Postprocessors

I documented the logic well with comments so intentions should be clear.

I also added these autocompletion scenarios to the MooseServer unit test.

Could I ask for a code review to get this merged for the NEAMS milestone?

**Comment by brandonlangley** [comment]
@dschwen @loganharbour -

Thanks a lot for the update in https://github.com/idaholab/moose/pull/25027 that removes the `_type` parameter.

I'll rebase this branch on that change in the next day or so, and I'll also simplify this [condition](https://github.com/brandonlangley/moose/blob/moose_server_autocomplete_22766/framework/src/base/MooseServer.C#L460)

```cpp
if (deprecated || is_private || existing_params.count(param_name) || param_name == "_type")
  continue;
```

in my autocomplete parameter filtering to just be the following, which will now be equivalent.

```cpp
if (deprecated || is_private || existing_params.count(param_name))
  continue;
```

Additionally, I just made this video of Workbench using the language server from this branch for autocompletion.

It doesn't include all features added here or anything but should give an idea of how some of this logic plays out.

In the bottom validation panel, you'll also see the input diagnostics that the server reports as the file is modified.

https://github.com/idaholab/moose/assets/21372118/f5c86fad-dc27-4a8b-876b-8e25269a0245

**Comment by brandonlangley** [comment]
@dschwen @loganharbour -

I rebased this to include the `_type` removal from https://github.com/idaholab/moose/pull/25027 and also removed that condition from my autocomplete filtering.

Let me know if you have any issues or concerns with this PR or if you're slammed and I should ask someone else to review.

File: unit/src/MooseServerTest.C [hunk 1]
```diff
@@ -663,7 +1141,164 @@ TEST_F(MooseServerTest, DocumentCompletionRequest)
...
+                                                   response_errors,
+                                                   response_id,
+                                                   response_is_incomplete,
+                                                   completions_array));
+
+  EXPECT_TRUE(response_errors.str().empty());
+
+  EXPECT_EQ(request_id, response_id);
+
+  EXPECT_EQ(4, (int)completions_array.size());
```
Comment (dschwen) [inline unit/src/MooseServerTest.C:1286]:
```suggestion
  EXPECT_EQ(4u, completions_array.size());
```

**Review by dschwen — COMMENTED** [review]
This looks mostly fine. I have some minor comments that you may wish to dismiss as they are mostly cosmetic :-)

**Comment by brandonlangley** [comment]
@dschwen -

Thanks a lot for the review feedback and suggestions, I just pushed a [commit](https://github.com/idaholab/moose/pull/24927/commits/145e8369fd68a721a37a5812b1a4bc43f0d3d3b7) that addresses all of your comments.

I applied those same style changes to the rest of the MooseServer code as well, which removed a lot of blank lines.

I agree that clustering sections of logic like this and using blank lines more sparingly does improve code readability.

I'm going to keep these notes in mind moving forward and continue this style when writing any future MOOSE code.

- - - - -

I added one more small [commit](https://github.com/idaholab/moose/pull/24927/commits/7bb7d0c308c4fc6eb661f7a77590716480fc2927) that fixes an issue observed if the MooseServer app tries to parse an empty string.

The Parser's optional text string parameter was defaulted as empty which was meant to indicate it wasn't provided.

This didn't work if a client passed the server an actual empty input as the Parser would fall back to parse from disk.

Strings don't have a natural sentinel value indicating absence, so I fixed this by making the parameter std::optional.

- - - - -

I went ahead and rebased these changes on top of devel and I think this should be good for final review and merge.

If you get a chance could you take a look at these two newest commits and let me know if you approve everything?

File: framework/include/base/MooseServer.h [hunk 2]
```diff
@@ -68,6 +72,99 @@ class MooseServer : public wasp::lsp::ServerImpl
                                      int line,
                                      int character);
 
+  /** get names of parameters and subblocks specified in given input node
```
[the remark this answers is not in the document]
Comment (brandonlangley) [reply framework/include/base/MooseServer.h]: Thanks, I pushed a commit that address this.

File: framework/src/base/MooseServer.C [hunk 3]
```diff
@@ -168,18 +188,675 @@ MooseServer::updateDocumentTextChanges(const std::string & replacement_text,
...
+
+  // get set of all parameter and subblock names already specified in input
+
+  std::set<std::string> existing_params, existing_subblocks;
+
+  getExistingInput(object_context, existing_params, existing_subblocks);
+
+  // set used to gather all parameters valid from object context of request
+
+  InputParameters valid_params = emptyInputParameters();
```
[the remark this answers is not in the document]
Comment (brandonlangley) [reply framework/src/base/MooseServer.C]: Thanks, I pushed a commit that address this.

File: framework/src/base/MooseServer.C [hunk 4]
```diff
@@ -168,18 +188,675 @@ MooseServer::updateDocumentTextChanges(const std::string & replacement_text,
...
+      std::string syntax_path = "/" + syntax_path_iter.first;
+
+      // skip current syntax path if already processed for different action
+
+      if (!syntax_paths_processed.insert(syntax_path).second)
+        continue;
+
+      // walk backward through syntax path adding subblock names to parents
+
+      for (size_t last_slash; (last_slash = syntax_path.find_last_of("/")) != std::string::npos;)
```
[the remark this answers is not in the document]
Comment (brandonlangley) [reply framework/src/base/MooseServer.C]: Thanks, I pushed a commit that address this.

File: unit/src/MooseServerTest.C [hunk 5]
```diff
@@ -83,6 +83,96 @@ protected:
...
+      std::string new_text;
+      std::string documentation;
+      int start_line;
+      int start_character;
+      int end_line;
+      int end_character;
+    };
+
+    std::vector<CompletionInfo> completions;
+    size_t max_label = 0;
```
[the remark this answers is not in the document]
Comment (brandonlangley) [reply unit/src/MooseServerTest.C]: Thanks, I pushed a commit that address this.

File: unit/src/MooseServerTest.C (hunk 1 above)
Comment (brandonlangley) [reply unit/src/MooseServerTest.C]: Thanks, I pushed a commit that address this.

**Review by loganharbour — APPROVED** [review]

**Comment by loganharbour** [comment]
I've verified that @dschwen's comments are addressed.

The logic with `optional` looks better.

**Pull request merged**